File: client/lib/models/task.dart
import 'user.dart';

class Task {
  final String id;
  final String title;
  final String description;
  final String type; // 'Personal' or 'Group'
  final DateTime? dueDate;
  final bool isCompleted;
  final List<TaskAssignment> assignedTo;
  final User? creatorId;

  Task({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    this.dueDate,
    required this.isCompleted,
    required this.assignedTo,
    this.creatorId,
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['_id'],
      title: json['title'],
      description: json['description'] ?? '',
      type: json['type'],
      dueDate: json['dueDate'] != null ? DateTime.parse(json['dueDate']) : null,
      isCompleted: json['isCompleted'] ?? false,
      assignedTo: (json['assignedTo'] as List<dynamic>?)
              ?.map((e) => TaskAssignment.fromJson(e))
              .toList() ??
          [],
      creatorId: json['creatorId'] != null && json['creatorId'] is Map<String, dynamic>
          ? User.fromJson(json['creatorId'])
          : null,
    );
  }
}

class TaskAssignment {
  final User? user;
  final String status; // 'Pending', 'Accepted', 'Rejected'
  final bool isCompleted;

  TaskAssignment({this.user, required this.status, required this.isCompleted});

  factory TaskAssignment.fromJson(Map<String, dynamic> json) {
    return TaskAssignment(
      user: json['userId'] != null && json['userId'] is Map<String, dynamic>
          ? User.fromJson(json['userId'])
          : null,
      status: json['status'],
      isCompleted: json['isCompleted'] ?? false,
    );
  }
}

File: server/models/Task.js
const mongoose = require('mongoose');

const taskSchema = new mongoose.Schema({
    creatorId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    title: { type: String, required: true },
    description: { type: String },
    type: { type: String, enum: ['Personal', 'Group'], required: true },
    dueDate: { type: Date },
    isCompleted: { type: Boolean, default: false },

    // For Group Tasks
    assignedTo: [{
        userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
        status: { type: String, enum: ['Pending', 'Accepted', 'Rejected'], default: 'Pending' },
        isCompleted: { type: Boolean, default: false }
    }]
}, { timestamps: true });

module.exports = mongoose.model('Task', taskSchema);
